torch/csrc/autograd/FunctionsManual.cpp

```diff
@@ -1416,30 +1416,44 @@ Tensor mm_mat1_sparse_backward(
       mat2.layout());
 }
 
-Tensor sparse_mask_like_grad(const Tensor& x, const Tensor& gx) {
+Tensor sparse_mask_like_grad(
+    const Tensor& x,
+    const Tensor& gx,
+    bool accumulate_matches) {
   if (x.is_coalesced() && gx.is_coalesced()) {
     if (x._nnz() >= gx._nnz()) {
       // search into x is faster
-      return gx._sparse_mask_projection(x);
+      return gx._sparse_mask_projection(x, accumulate_matches);
     } else {
       // search into gx is faster
       return gx.sparse_mask(x);
     }
   } else if (x.is_coalesced()) {
     return gx.sparse_mask(x);
   } else if (gx.is_coalesced()) {
-    return gx._sparse_mask_projection(x);
+    return gx._sparse_mask_projection(x, accumulate_matches);
   } else {
     if (x._nnz() >= gx._nnz()) {
       // gx.coalesce() is likely faster
-      return gx.coalesce()._sparse_mask_projection(x);
+      return gx.coalesce()._sparse_mask_projection(x, accumulate_matches);
     } else {
       // x.coalesce() is likely faster
       return gx.sparse_mask(x.coalesce());
     }
   }
 }
 
+Tensor sparse_mask_backward(
+    const Tensor& grad,
+    const Tensor& mask,
+    const c10::Layout self_layout) {
+  // NOTE: sparse_mask accumulates matches, so the backward step has to
+  // accumulate as well.
+  const auto self_grad =
+      sparse_mask_like_grad(mask, grad, /*accumulate_matches=*/true);
+  return self_layout == at::kStrided ? self_grad.to_dense() : self_grad;
```

Comment:
Why do we have to convert `self_grad` tensor to a strided tensor when `self` is strided?

Reply:
Because autograd does not like layout mismatch? Or was it fixed already?

Reply:
An answer is likely provided by applying:
```suggestion
  return self_grad;
```
locally. It either works or you can report an error message here to resolve this conversation.

Reply:
Had to revert, looks like not all CIs are up to date with the change...

Reply:
It looks like gradcheck is failing:
```
torch.autograd.gradcheck.GradcheckError: grad is incorrect layout (torch.sparse_coo is not torch.strided)
```
which by itself does not mean that returning non-strided grad could not be supported. We just have to relax gradcheck and see if the non-gradcheck parts of pytorch are fine with it. But not in this PR...